### PR TITLE
Fix SFT for base models

### PR DESCRIPTION
**Pull request by lewtun**
Fixes https://github.com/huggingface/open-r1/issues/603 https://github.com/huggingface/open-r1/issues/556 https://github.com/huggingface/open-r1/issues/520 https://github.com/huggingface/open-r1/issues/515 https://github.com/huggingface/open-r1/pull/494

This PR fixes several issues associated **unbounded generation** after applying SFT on **base** models. The reason we never ran into this internally is because we have thus far applied SFT on **instruct** models where (a) a chat template is defined and (b) the EOS token is aligned with the chat template.

Below I list out the 3 main cases I saw in various issues and provide the corresponding solution. Note that one must bump `trl` to make use of these fixes.

## Case 1: base model has a pre-defined chat template
Models like`Qwen/Qwen2.5-1.5B` have a [pre-defined chat template](https://huggingface.co/Qwen/Qwen2.5-1.5B/blob/main/tokenizer_config.json#L198) (ChatML). As a result, running SFT on Qwen models would format the `messages` in ChatML (as expected), but the EOS token of the SFT model was misaligned: it remained the same as the base model's `<|endoftext|>` instead of ChatML's `<|im_end|>`. 

### Solution:

Specify the chat template's EOS token as follows:

```diff
ACCELERATE_LOG_LEVEL=info TRANSFORMERS_VERBOSITY=info accelerate launch --config_file recipes/accelerate_configs/zero2.yaml src/open_r1/sft.py \
    --model_name_or_path Qwen/Qwen2.5-1.5B \
+   --eos_token '<|im_end|>' \
    --dataset_name trl-lib/Capybara \
    --learning_rate 2.0e-5 \
    --max_seq_length 2048 \
    --num_train_epochs 2 \
    --per_device_train_batch_size 16 \
    --gradient_accumulation_steps 1 \
    --gradient_checkpointing \
    --bf16 \
    --use_liger_kernel \
    --logging_steps 5 \
    --eval_strategy steps \
    --eval_steps 100 \
    --output_dir data/Qwen2.5-1.5B-SFT-Capybara-No-Packing \
    --push_to_hub
```

**Model:** https://huggingface.co/lewtun/Qwen2.5-1.5B-SFT-Capybara-No-Packing

Test inference with:

```shell
transformers-cli chat --model_name_or_path lewtun/Qwen2.5-1.5B-SFT-Capybara-No-Packing --do_sample --max_new_tokens 512
```

## Case 2: no chat template is specified

Models like `meta-llama/Llama-3.2-1B-Instruct` have no pre-defined chat template (the norm for most base models). As a result, unless the user set the `--chat_template` arg manually, the model would be trained without a chat template. 

### Solution
We now use `setup_chat_format()` which defaults to ChatML if the model does not have a chat template defined.

```shelll
ACCELERATE_LOG_LEVEL=info TRANSFORMERS_VERBOSITY=info accelerate launch --config_file recipes/accelerate_configs/zero2.yaml src/open_r1/sft.py \
    --model_name_or_path meta-llama/Llama-3.2-1B \
    --dataset_name trl-lib/Capybara \
    --learning_rate 2.0e-5 \
    --max_seq_length 2048 \
    --num_train_epochs 2 \
    --per_device_train_batch_size 16 \
    --gradient_accumulation_steps 1 \
    --gradient_checkpointing \
    --bf16 \
    --use_liger_kernel \
    --logging_steps 5 \
    --eval_strategy steps \
    --eval_steps 100 \
    --output_dir data/Llama-3.2-1B-SFT-Capybara-No-Packing-ChatML \
    --push_to_hub
```

**Model:** https://huggingface.co/lewtun/Llama-3.2-1B-SFT-Capybara-No-Packing-ChatML

Test inference with:

```shell
transformers-cli chat --model_name_or_path lewtun/Llama-3.2-1B-SFT-Capybara-No-Packing-ChatML --do_sample --max_new_tokens 512
```

## Case 3: custom chat template specified
If a user wished to use a chat template like Llama's, a similar issue as Case 1 occurs: there is a misalignment between the base model's EOS token `<|end_of_text|>` and the chat template's end of turn `<|eot_id|>`

### Solution
The user provides both the chat template and corresponding EOS token

```diff
ACCELERATE_LOG_LEVEL=info TRANSFORMERS_VERBOSITY=info accelerate launch --config_file recipes/accelerate_configs/zero2.yaml src/open_r1/sft.py \
    --model_name_or_path meta-llama/Llama-3.2-1B \
+   --chat_template "$(cat scratch/llama_chat_template.jinja)" \
+   --eos_token '<|eot_id|>' \
    --dataset_name trl-lib/Capybara \
    --learning_rate 2.0e-5 \
    --max_seq_length 2048 \
    --num_train_epochs 2 \
    --per_device_train_batch_size 16 \
    --gradient_accumulation_steps 1 \
    --gradient_checkpointing \
    --bf16 \
    --use_liger_kernel \
    --logging_steps 5 \
    --eval_strategy steps \
    --eval_steps 100 \
    --output_dir data/Llama-3.2-1B-SFT-Capybara-No-Packing-Llama \
    --push_to_hub
```

**Model** https://huggingface.co/lewtun/Llama-3.2-1B-SFT-Capybara-No-Packing-Llama

Test inference with:

```shell
transformers-cli chat --model_name_or_path lewtun/Llama-3.2-1B-SFT-Capybara-No-Packing-Llama --do_sample --max_new_tokens 512
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,7 +112,6 @@ accelerate launch --config_file=recipes/accelerate_configs/zero3.yaml src/open_r
     --dataset_name open-r1/OpenR1-Math-220k \
     --learning_rate 5.0e-5 \
     --num_train_epochs 1 \
-    --packing \
```
Comment (lewtun) [inline README.md:115]: Removing packing as the default recommendation since we found a slight performance degradation with reasoning models (data gets split across chunk boundaries)

**Comment by BigWavelet** [comment]
Thanks so much. does it works for fine-tuning on qwen-insutrct model with openr1-220k-math?

**Comment by BigWavelet** [comment]
> Thanks so much. does it works for fine-tuning on qwen-insutrct model with openr1-220k-math?

I mean sft on qwen2.5-1.5b-instruct  with openr1-220k-math.

**Review by edbeeching — APPROVED** [review]

**Comment by BigWavelet** [comment]
After sft on qwen2.5-1.5b-instruct with openr1-220k-math, I also got repeation output on gsm8k, do you have any ideas? Thx.

**Comment by lewtun** [comment]
> After sft on qwen2.5-1.5b-instruct with openr1-220k-math, I also got repeation output on gsm8k, do you have any ideas? Thx.

Hey @BigWavelet here's a Qwen base model I trained on `open-r1/OpenR1-Math-220k` using : https://huggingface.co/lewtun/Qwen2.5-1.5B-Open-R1-Distill

It seems to work fine in my vibe testing:

```shell
transformers-cli chat --model_name_or_path lewtun/Qwen2.5-1.5B-Open-R1-Distill --do_sample --max_new_tokens 4096
```

I trained it as follows (note you need to pass the EOS token for Qwen base models):

```shell
accelerate launch --config_file recipes/accelerate_configs/zero2.yaml src/open_r1/sft.py \
     --config recipes/Qwen2.5-1.5B-Instruct/sft/config_demo.yaml \
     --model_name_or_path Qwen/Qwen2.5-1.5B \
     --eos_token '<|im_end|>'
```

**Comment by ItGirls** [comment]
problem not solved, we got several <think> <answer> tags when use our trained DeepSeek-R1-Distill-Qwen-7B model. And I checked the pad token and eos token,  they  are the same and also EOS token is aligned with the chat template. How to solve this problem

**Comment by lewtun** [comment]
Hello @ItGirls can you share the command you used to train with? For fine-tuning a base model, you should ideally have an EOS token that differs from the padding one, like here: https://huggingface.co/lewtun/Qwen2.5-1.5B-Open-R1-Distill/blob/main/tokenizer_config.json#L200

**Comment by ItGirls** [comment]
> For fine-tuning a base model, you should ideally have an EOS token that differs from the padding one
accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/sft.py \
    --config recipes/ape/sft/config.yaml
config.yaml:
# Model arguments
model_name_or_path: /xxx/deepseek-ai/DeepSeek-R1-Distill-Qwen-7B
model_revision: main
torch_dtype: bfloat16
attn_implementation: flash_attention_2

# Data training arguments
dataset_name: /xxx/xxx

# TODO DeepSeek-R1-Distill-Qwen-7B自己的 (会丢弃assitant中的推理部分)，但是llama_factory与open_r1不一样，不会用tokenizer.apply_chat_template，所以不影响
# chat_template: "{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{% set ns = namespace(is_first=false, is_tool=false, is_output_first=true, system_prompt='') %}{%- for message in messages %}{%- if message['role'] == 'system' %}{% set ns.system_prompt = message['content'] %}{%- endif %}{%- endfor %}{{bos_token}}{{ns.system_prompt}}{%- for message in messages %}{%- if message['role'] == 'user' %}{%- set ns.is_tool = false -%}{{'<｜User｜>' + message['content']}}{%- endif %}{%- if message['role'] == 'assistant' and message['content'] is none %}{%- set ns.is_tool = false -%}{%- for tool in message['tool_calls']%}{%- if not ns.is_first %}{{'<｜Assistant｜><｜tool▁calls▁begin｜><｜tool▁call▁begin｜>' + tool['type'] + '<｜tool▁sep｜>' + tool['function']['name'] + '\\n' + '```json' + '\\n' + tool['function']['arguments'] + '\\n' + '```' + '<｜tool▁call▁end｜>'}}{%- set ns.is_first = true -%}{%- else %}{{'\\n' + '<｜tool▁call▁begin｜>' + tool['type'] + '<｜tool▁sep｜>' + tool['function']['name'] + '\\n' + '```json' + '\\n' + tool['function']['arguments'] + '\\n' + '```' + '<｜tool▁call▁end｜>'}}{{'<｜tool▁calls▁end｜><｜end▁of▁sentence｜>'}}{%- endif %}{%- endfor %}{%- endif %}{%- if message['role'] == 'assistant' and message['content'] is not none %}{%- if ns.is_tool %}{{'<｜tool▁outputs▁end｜>' + message['content'] + '<｜end▁of▁sentence｜>'}}{%- set ns.is_tool = false -%}{%- else %}{% set content = message['content'] %}{% if '</think>' in content %}{% set content = content.split('</think>')[-1] %}{% endif %}{{'<｜Assistant｜>' + content + '<｜end▁of▁sentence｜>'}}{%- endif %}{%- endif %}{%- if message['role'] == 'tool' %}{%- set ns.is_tool = true -%}{%- if ns.is_output_first %}{{'<｜tool▁outputs▁begin｜><｜tool▁output▁begin｜>' + message['content'] + '<｜tool▁output▁end｜>'}}{%- set ns.is_output_first = false %}{%- else %}{{'\\n<｜tool▁output▁begin｜>' + message['content'] + '<｜tool▁output▁end｜>'}}{%- endif %}{%- endif %}{%- endfor -%}{% if ns.is_tool %}{{'<｜tool▁outputs▁end｜>'}}{% endif %}{% if add_generation_prompt and not ns.is_tool %}{{'<｜Assistant｜>'}}{% endif %}"
# 修改过的chat_template 
chat_template: "{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{% set ns = namespace(is_first=false, is_tool=false, is_output_first=true, system_prompt='') %}{%- for message in messages %}{%- if message['role'] == 'system' %}{% set ns.system_prompt = message['content'] %}{%- endif %}{%- endfor %}{{bos_token}}{{ns.system_prompt}}{%- for message in messages %}{%- if message['role'] == 'user' %}{%- set ns.is_tool = false -%}{{'<｜User｜>' + message['content']}}{%- endif %}{%- if message['role'] == 'assistant' and message['content'] is none %}{%- set ns.is_tool = false -%}{%- for tool in message['tool_calls']%}{%- if not ns.is_first %}{{'<｜Assistant｜><｜tool▁calls▁begin｜><｜tool▁call▁begin｜>' + tool['type'] + '<｜tool▁sep｜>' + tool['function']['name'] + '\\n' + '```json' + '\\n' + tool['function']['arguments'] + '\\n' + '```' + '<｜tool▁call▁end｜>'}}{%- set ns.is_first = true -%}{%- else %}{{'\\n' + '<｜tool▁call▁begin｜>' + tool['type'] + '<｜tool▁sep｜>' + tool['function']['name'] + '\\n' + '```json' + '\\n' + tool['function']['arguments'] + '\\n' + '```' + '<｜tool▁call▁end｜>'}}{{'<｜tool▁calls▁end｜><｜end▁of▁sentence｜>'}}{%- endif %}{%- endfor %}{%- endif %}{%- if message['role'] == 'assistant' and message['content'] is not none %}{%- if ns.is_tool %}{{'<｜tool▁outputs▁end｜>' + message['content'] + '<｜end▁of▁sentence｜>'}}{%- set ns.is_tool = false -%}{%- else %}{% set content = message['content'] %}{{'<｜Assistant｜>' + content + '<｜end▁of▁sentence｜>'}}{%- endif %}{%- endif %}{%- if message['role'] == 'tool' %}{%- set ns.is_tool = true -%}{%- if ns.is_output_first %}{{'<｜tool▁outputs▁begin｜><｜tool▁output▁begin｜>' + message['content'] + '<｜tool▁output▁end｜>'}}{%- set ns.is_output_first = false %}{%- else %}{{'\\n<｜tool▁output▁begin｜>' + message['content'] + '<｜tool▁output▁end｜>'}}{%- endif %}{%- endif %}{%- endfor -%}{% if ns.is_tool %}{{'<｜tool▁outputs▁end｜>'}}{% endif %}{% if add_generation_prompt and not ns.is_tool %}{{'<｜Assistant｜>'}}{% endif %}"
system_prompt: "xxx。请按以下格式回答: <think>\n...\n</think>\n<answer>\n...\n</answer>"
dataset_num_proc: 48

#SFT hyperparam
max_length: 23964
weight_decay: 0.0001
optim: adamw_torch
lr_scheduler_type: linear
warmup_ratio: 0.1
learning_rate: 5.0e-05
gradient_accumulation_steps: 8 #1
per_device_eval_batch_size: 1
per_device_train_batch_size: 1

# SFT trainer config
max_steps: -1
num_train_epochs: 1 #20
bf16: true
do_eval: false
use_liger_kernel: true
eval_strategy: 'no'
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
hub_model_id: OpenR1-Qwen-7B-SFT
hub_strategy: every_save
log_level: info
logging_steps: 1
logging_strategy: steps
packing: true
output_dir: data/Qwen2.5-7B-Open-R1-Distill-0430
overwrite_output_dir: false
push_to_hub: false
report_to:
- wandb
save_strategy: "steps"
save_steps: 120
save_total_limit: 1
seed: 42
----
" For fine-tuning a base model, you should ideally have an EOS token that differs from the padding one", in chat_template, the eos token is <｜end▁of▁sentence｜>.If I want to set a new eos token, I also need to change the eos token in the chat_template,right?

**Comment by Hasuer** [comment]
> problem not solved, we got several <think> <answer> tags when use our trained model based on DeepSeek-R1-Distill-Qwen-7B model.
> 
> And I checked the pad token and eos token, they are the same and also EOS token is aligned with the chat template(see here [DeepSeek-R1-Distill-Qwen-7B](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-7B/blob/main/tokenizer_config.json) How to solve this problem.

Hi, have you solved the problem?
